File: src_backup/main/java/com/myudog/myulib/internal/state/ParticleState.java
package com.myudog.myulib.internal.state;
import net.minecraft.world.phys.Vec3;public final class ParticleState {;
    public Vec3 pos;
    public Vec3 vel;
    public int age;
    public final int maxAge;
    public final double randomSeed;
    public ParticleState(Vec3 pos, Vec3 vel, int age, int maxAge) {
        this(pos, vel, age, maxAge, Math.random());
    }
    public ParticleState(Vec3 pos, Vec3 vel, int age, int maxAge, double randomSeed) {
        this.pos = pos;
        this.vel = vel;
        this.age = age;
        this.maxAge = maxAge;
        this.randomSeed = randomSeed;
    }
    public float getProgress() {
        if (maxAge <= 0) {
            return 1.0f;
        }
        return Math.min(1.0f, age / (float) maxAge);
    }
    public boolean isDead() {
        return age >= maxAge;
    }
}

File: src/main/kotlin/com/myudog/myulib/internal/state/ParticleState.kt
package com.myudog.myulib.internal.state

import net.minecraft.util.math.Vec3d

internal data class ParticleState(
    var pos: Vec3d,           // 相對座標
    var vel: Vec3d,           // 速度向量
    var age: Int = 0,         // 當前歲數 (ticks)
    val maxAge: Int,          // 最大壽命
    val randomSeed: Double = Math.random() // 用於個別粒子的隨機行為
) {
    val progress: Float get() = age.toFloat() / maxAge // 生命進度 (0.0 ~ 1.0)
    val isDead: Boolean get() = age >= maxAge
}